prow/spyglass/viewers/buildlog/buildlog_viewer.go

```diff
@@ -0,0 +1,273 @@
+		ViewName:     name,
+	}
+	switch currentViewData.Operation {
+	case "all":
+		logLines = logLinesAll(artifact)
+		newArtifactView.ViewMethodDescription = fmt.Sprintf("viewing all lines")
+		newArtifactView.ViewAll = true
+	case "more":
+		if currentViewData.UseBytes {
+			nBytes := int64(currentViewData.HeadBytes + byteChunkSize)
+			artifactSize, err := artifact.Size()
+			if err != nil {
+				logrus.WithFields(logrus.Fields{"artifact": artifact.JobPath()}).WithError(err).Error("Reading log lines failed.")
+			}
+			if nBytes >= artifactSize {
+				logLines = logLinesAll(artifact)
+				newArtifactView.ViewMethodDescription = fmt.Sprintf("viewing all lines")
+				newArtifactView.ViewAll = true
+			} else {
+				newArtifactView.ViewMethodDescription = fmt.Sprintf("viewing first %d bytes", nBytes)
+				newLogViewData.HeadBytes = nBytes
+				newLogViewData.UseBytes = true
+				logLines = logLinesHead(artifact, nBytes)
+			}
+		} else {
+			nLines := currentViewData.TailLines + lineChunkSize
+			logLines, err = viewers.LastNLines(artifact, nLines)
+			if err != nil {
+				if err == viewers.ErrGzipOffsetRead { // Try a different read operation
+					nBytes := int64(byteChunkSize)
+					newLogViewData.UseBytes = true
+					newLogViewData.HeadBytes = nBytes
+					newArtifactView.ViewMethodDescription = fmt.Sprintf("viewing first %d bytes", nBytes)
+					logLines = logLinesHead(artifact, nBytes)
+				} else {
+					logrus.WithError(err).Error("Reading log lines failed.")
+					logLines = []string{}
+				}
+			} else {
+				newLogViewData.TailLines = nLines
+				newArtifactView.ViewMethodDescription = fmt.Sprintf("viewing last %d lines", nLines)
+			}
+		}
+	case "":
+	default:
+	}
+
+	newArtifactView.LogLines = logLines
+	return newArtifactView, newLogViewData
+
+}
+
+// logLinesHead reads the first n bytes of an artifact and splits it into lines.
+func logLinesHead(artifact viewers.Artifact, n int64) []string {
+	read, err := artifact.ReadAtMost(n)
+	if err != nil {
+		if err == io.ErrUnexpectedEOF {
+			artifactSize, err := artifact.Size()
+			if err != nil {
+				logrus.WithError(err).WithFields(logrus.Fields{
+					"n":        strconv.FormatInt(n, 10),
+					"artifact": artifact.JobPath(),
+				}).Error("Unexpected EOF, failed to get size of artifact.")
+			} else {
+				logrus.WithError(err).WithFields(logrus.Fields{
+					"n":            strconv.FormatInt(n, 10),
+					"artifactSize": strconv.FormatInt(artifactSize, 10),
+				}).Info("Unexpected EOF, continuing...")
+			}
+		}
+	}
+	logLines := strings.Split(string(read), "\n")
+	for ix, line := range logLines {
+		logLines[ix] = fmt.Sprintf("%d.\t%s", ix+1, line)
+	}
+	return logLines
+}
+
+// logLinesAll reads all of an artifact and splits it into lines.
+func logLinesAll(artifact viewers.Artifact) []string {
+	read, err := artifact.ReadAll()
+	if err != nil {
+		if err == viewers.ErrFileTooLarge {
+			logrus.WithError(err).Error("Artifact too large to read all.")
+			return []string{}
```

Comment:
Seems like we need a mechanism to pass errors up, this is weird

Reply:
Pass errors up? If the viewers find they are unable to deal with a certain artifact they can just return a div with the error no?